### ISSUE TITLE
Problems with limited inner buffers for BoundaryReader. 

### ISSUE THREAD

**Issue by nicolas-cherel** [cybergeek94/multipart#18]
I'm not 100% percent sure what's happening here, since I can't reproduce the bug reliably, it happens on a hyper server requests handling, but sometimes [`read_to_boundary()`](https://github.com/cybergeek94/multipart/blob/202f5e0418b7b59591862a9d577af657d666d8eb/src/server/boundary.rs#L30-L54) fails to read boundary properly. 

For the cases I managed to inspect, it was when the boundary was not entirely in the buffer, and truncated, with the boundary end visible on the next `fill_buf()`. From what I see, the code cannot handle this case, but I'm not sure if the bug is not elsewhere, and that the buffer should be large enough to fit the boundary every time. I tried to force buffer capacity with the given content length of the request, but it did not fix it, it's pretty confusing.

**Comment by cybergeek94** [comment]
I thought I had covered this, I guess not. `BoundaryReader` is definitely supposed to be able to handle this. It's probably just a logic error in `read_to_boundary()`, we've caught plenty of those already.

**Comment by cybergeek94** [comment]
I have a partial fix on the `boundary_fix` branch. I added a test to make sure it handles a split on the boundary correctly.

It ended up being necessary to have `BoundaryReader` be aware of the `--` after the last boundary. Otherwise, `Multipart` would keep reading an empty string looking for it, since `BoundaryReader` will never yield bytes that *might* be part of a boundary.

However, `test_split_boundary` gets stuck in an infinite loop at the consume at https://github.com/cybergeek94/multipart/blob/boundary_fix/src/server/boundary.rs#L235 and I'm at a bit of a loss. I'm still working on it, though.

**Comment by nicolas-cherel** [comment]
Great, thanks a lot !

Ho, the infinite loop does ring a bell, there is changes in your branch, but [here](https://github.com/cybergeek94/multipart/blob/boundary_fix/src/server/boundary.rs#L65-L68), there might be cases where the state of `self.boundary_read` or `self.at_end` is not changed, while returned buffer is empty, with nothing left to be read.

**Comment by cybergeek94** [comment]
That fixed the infinite loop but the test still fails. Interestingly enough, it looks to be right at the split point set by the magic number in `test_split_boundary()` but I'm not sure yet why that would be the case.

**Comment by cybergeek94** [comment]
Alas, I think I've run into the same limitation of `BufRead`/`BufReader`'s API that bit me in the butt the last time I tried to use it. The problem is, `BufRead::fill_buf()` will not read from the underlying stream until all bytes currently in the buffer are consumed. 

This is a *really* big problem because we don't want to consume bytes from the buffer unless we know it is safe to do so; often, the bytes we'd have to consume are actually part of a boundary, because otherwise we'd just yield them to the user. 

There's only two ways we could fix this: read into a temporary buffer to force `BufReader` to pull in more bytes (excess allocations, code bloat), or reimplement `BufReader` ourselves (just... bad all around). I tried both approaches before and neither were satisfactory.
